libbpf-tools/biotop.c

```diff
@@ -0,0 +1,438 @@
+
+	while (1) {
+		err = bpf_map_get_next_key(fd, prev_key, &datas[rows].key);
+		if (err) {
+			if (errno == ENOENT) {
+				err = 0;
+				break;
+			}
+			warn("bpf_map_get_next_key failed: %s\n", strerror(errno));
+			return err;
+		}
+		err = bpf_map_lookup_elem(fd, &datas[rows].key, &datas[rows].value);
+		if (err) {
+			warn("bpf_map_lookup_elem failed: %s\n", strerror(errno));
+			return err;
+		}
+		prev_key = &datas[rows].key;
+		rows++;
+	}
+
+	qsort(datas, rows, sizeof(struct data_t), sort_column);
+	rows = rows < output_rows ? rows : output_rows;
+	for (i = 0; i < rows; i++) {
+		int major;
+		int minor;
+		struct info_t *key = &datas[i].key;
+		struct val_t *value = &datas[i].value;
+		float avg_ms = 0;
+
+		/* To avoid floating point exception. */
+		if (value->io)
+			avg_ms = ((float) value->us) / 1000 / value->io;
+
+		major = key->major;
+		minor = key->minor;
+
+		printf("%-6d %-16s %1s %-3d %-3d %-8s %5d %7lld %6.2f\n",
+		       key->pid, key->name, key->rwflag ? "W": "R",
+		       major, minor, search_disk_name(major, minor),
+		       value->io, value->bytes / 1024, avg_ms);
+	}
+
+	printf("\n");
+	prev_key = NULL;
+
+	while (1) {
+		struct info_t key;
+
+		err = bpf_map_get_next_key(fd, prev_key, &key);
+		if (err) {
+			if (errno == ENOENT) {
+				err = 0;
+				break;
+			}
+			warn("bpf_map_get_next_key failed: %s\n", strerror(errno));
+			return err;
+		}
+		err = bpf_map_delete_elem(fd, &key);
+		if (err) {
+			warn("bpf_map_delete_elem failed: %s\n", strerror(errno));
+			return err;
+		}
+		prev_key = &key;
+	}
+	return err;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct biotop_bpf *obj;
+	struct ksyms *ksyms;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_strict_mode(LIBBPF_STRICT_ALL);
+	libbpf_set_print(libbpf_print_fn);
+
+	obj = biotop_bpf__open();
+	if (!obj) {
+		warn("failed to open BPF object\n");
+		return 1;
+	}
+
+	parse_disk_stat();
+
+	ksyms = ksyms__load();
+	if (!ksyms) {
+		err = -ENOMEM;
+		warn("failed to load kallsyms\n");
+		goto cleanup;
+	}
+
+	err = biotop_bpf__load(obj);
+	if (err) {
+		warn("failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	if (ksyms__get_symbol(ksyms, "__blk_account_io_start"))
+		bpf_program__attach_kprobe(obj->progs.blk_account_io_start, false, "__blk_account_io_start");
+	if (ksyms__get_symbol(ksyms, "blk_account_io_start"))
+		bpf_program__attach_kprobe(obj->progs.blk_account_io_start, false, "blk_account_io_start");
+
+	if (ksyms__get_symbol(ksyms, "__blk_account_io_done"))
+		bpf_program__attach_kprobe(obj->progs.blk_account_io_done, false, "__blk_account_io_done");
+	if (ksyms__get_symbol(ksyms, "blk_account_io_done"))
+		bpf_program__attach_kprobe(obj->progs.blk_account_io_done, false, "blk_account_io_done");
+
+	err = biotop_bpf__attach(obj);
```

Comment:
With `bpf_program__attach_kprobe`, you don't need `biotop_bpf__attach` any more.

You should check result of `bpf_program__attach_kprobe` and free the resources on exit.

For simplicity, you can have four dummy BPF programs, so the generated skeleton will have four bpf_link s, assign the attach results to these links and call `biotop_bpf__destroy` on exit, done.

Reply:
@eiffel-fl  This looks incorrect to me. Please fix.

Reply:
```diff
diff --git a/libbpf-tools/biotop.bpf.c b/libbpf-tools/biotop.bpf.c
index 0d418cd5..226e32d3 100644
--- a/libbpf-tools/biotop.bpf.c
+++ b/libbpf-tools/biotop.bpf.c
@@ -9,7 +9,6 @@
 #include "maps.bpf.h"
 #include "core_fixes.bpf.h"
 
-
 struct {
 	__uint(type, BPF_MAP_TYPE_HASH);
 	__uint(max_entries, 10240);
@@ -31,7 +30,7 @@ struct {
 	__type(value, struct val_t);
 } counts SEC(".maps");
 
-SEC("kprobe/blk_account_io_start")
+SEC("kprobe")
 int BPF_KPROBE(blk_account_io_start, struct request *req)
 {
 	struct who_t who = {};
@@ -57,7 +56,7 @@ int BPF_KPROBE(blk_mq_start_request, struct request *req)
 	return 0;
 }
 
-SEC("kprobe/blk_account_io_done")
+SEC("kprobe")
 int BPF_KPROBE(blk_account_io_done, struct request *req, u64 now)
 {
 	struct val_t *valp, zero = {};
diff --git a/libbpf-tools/biotop.c b/libbpf-tools/biotop.c
index 98f2fffd..2dc64658 100644
--- a/libbpf-tools/biotop.c
+++ b/libbpf-tools/biotop.c
@@ -7,6 +7,9 @@
  * Based on biotop(8) from BCC by Brendan Gregg.
  * 03-Mar-2022   Francis Laniel   Created this.
  */
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
 #include <argp.h>
 #include <errno.h>
 #include <signal.h>
@@ -70,7 +73,7 @@ void free_vector(struct vector vector) {
 	free(vector.elems);
 }
 
-struct vector disks = { 0 };
+struct vector disks = {};
 
 static volatile sig_atomic_t exiting = 0;
 
@@ -281,7 +284,7 @@ static int print_stat(struct biotop_bpf *obj)
 			printf("%8s loadavg: %s\n", ts, buf);
 		fclose(f);
 	}
-	printf("%-6s %-16s %1s %-3s %-3s %-8s %5s %7s %6s\n",
+	printf("%-7s %-16s %1s %-3s %-3s %-8s %5s %7s %6s\n",
 	       "PID", "COMM", "D", "MAJ", "MIN", "DISK", "I/O", "Kbytes", "AVGms");
 
 	while (1) {
@@ -319,7 +322,7 @@ static int print_stat(struct biotop_bpf *obj)
 		major = key->major;
 		minor = key->minor;
 
-		printf("%-6d %-16s %1s %-3d %-3d %-8s %5d %7lld %6.2f\n",
+		printf("%-7d %-16s %1s %-3d %-3d %-8s %5d %7lld %6.2f\n",
 		       key->pid, key->name, key->rwflag ? "W": "R",
 		       major, minor, search_disk_name(major, minor),
 		       value->io, value->bytes / 1024, avg_ms);
@@ -357,7 +360,6 @@ int main(int argc, char **argv)
 		.parser = parse_arg,
 		.doc = argp_program_doc,
 	};
-	struct bpf_link *probes[4] = { NULL };
 	struct biotop_bpf *obj;
 	struct ksyms *ksyms;
 	int err;
@@ -390,34 +392,22 @@ int main(int argc, char **argv)
 		goto cleanup;
 	}
 
-	if (ksyms__get_symbol(ksyms, "__blk_account_io_start")) {
-		probes[0] = bpf_program__attach_kprobe(obj->progs.blk_account_io_start, false, "__blk_account_io_start");
-		if (!probes[0]) {
-			warn("failed to load kprobe __blk_account_io_start\n");
-			goto cleanup;
-		}
-	}
-	if (ksyms__get_symbol(ksyms, "blk_account_io_start")) {
-		probes[1] = bpf_program__attach_kprobe(obj->progs.blk_account_io_start, false, "blk_account_io_start");
-		if (!probes[1]) {
-			warn("failed to load kprobe blk_account_io_start\n");
-			goto cleanup;
-		}
+	if (ksyms__get_symbol(ksyms, "__blk_account_io_start"))
+		obj->links.blk_account_io_start = bpf_program__attach_kprobe(obj->progs.blk_account_io_start, false, "__blk_account_io_start");
+	else
+		obj->links.blk_account_io_start = bpf_program__attach_kprobe(obj->progs.blk_account_io_start, false, "blk_account_io_start");
+	if (!obj->links.blk_account_io_start) {
+		warn("failed to attach blk_account_io_start\n");
+		goto cleanup;
 	}
 
-	if (ksyms__get_symbol(ksyms, "__blk_account_io_done")) {
-		probes[2] = bpf_program__attach_kprobe(obj->progs.blk_account_io_done, false, "__blk_account_io_done");
-		if (!probes[2]) {
-			warn("failed to load kprobe __blk_account_io_done\n");
-			goto cleanup;
-		}
-	}
-	if (ksyms__get_symbol(ksyms, "blk_account_io_done")) {
-		probes[3] = bpf_program__attach_kprobe(obj->progs.blk_account_io_done, false, "blk_account_io_done");
-		if (!probes[3]) {
-			warn("failed to load kprobe blk_account_io_done\n");
-			goto cleanup;
-		}
+	if (ksyms__get_symbol(ksyms, "__blk_account_io_done"))
+		obj->links.blk_account_io_done = bpf_program__attach_kprobe(obj->progs.blk_account_io_done, false, "__blk_account_io_done");
+	else
+		obj->links.blk_account_io_done = bpf_program__attach_kprobe(obj->progs.blk_account_io_done, false, "blk_account_io_done");
+	if (!obj->links.blk_account_io_done) {
+		warn("failed to attach blk_account_io_done\n");
+		goto cleanup;
 	}
 
 	err = biotop_bpf__attach(obj);
@@ -453,9 +443,6 @@ int main(int argc, char **argv)
 cleanup:
 	ksyms__free(ksyms);
 	free_vector(disks);
-	for (int i = 0; i < sizeof probes / sizeof probes[0]; i++)
-		if (probes[i])
-			bpf_link__destroy(probes[i]);
 	biotop_bpf__destroy(obj);
 
 	return err != 0;

```

Reply:
Try this small diff.

Reply:
Please paste the error here. By making `SEC("kprobe")`, we tell libbpf not to perform auto-attach, it should work with the latest libbpf submodule. Otherwise, calling `biotop_bpf__attach` will make them attach twice.